server/lib/nl/fulfillment/base.py

```diff
@@ -259,73 +270,91 @@ def _add_charts_with_place_fallback(state: PopulateState, places: List[Place],
 class ChartVarsExistenceCheckState:
   chart_vars: ChartVars
   # Existing svs from among chart_vars.svs
+  # Note that `chart_vars` is not mutated to point to existing SVs.
   exist_svs: List[str]
-  # Set only if chart_vars.event is set
+  # Set only if chart_vars.event is true, to indicate event existence.
   exist_event: bool = False
 
 
 @dataclass
 class SVExistenceCheckState:
   sv: str
   chart_vars_list: List[ChartVarsExistenceCheckState]
-
-  extended_svs: List[str]
-  # Existence check result for extended SVs.
-  extended_exist_svs: List[str]
+  # ChartVars for extended SVs.
+  extended_vars: ChartVarsExistenceCheckState
 
 
 #
 # This class helps batch existence checks.
 #
 class ExistenceCheckStateTracker:
 
+  # NOTE: If sv2extensions is set, then this is for extensions only.
   def __init__(self, state: PopulateState, places: List[str], svs: List[str],
                sv2extensions: Dict):
     self.state = state
     self.places = places
     self.all_svs = set()
     self.exist_sv_states: List[SVExistenceCheckState] = []
+    # Map of existing SVs with key as SV DCID and value as
+    # whether the SV has single-data point.
+    self.existing_svs = {}
 
     # Loop over all SVs, and construct existence check state.
     for rank, sv in enumerate(svs):
       exist_state = SVExistenceCheckState(sv=sv,
                                           chart_vars_list=[],
-                                          extended_svs=[],
-                                          extended_exist_svs=[])
-
-      chart_vars_list = _build_chart_vars(state, sv, rank)
-      for chart_vars in chart_vars_list:
-        exist_cv = ChartVarsExistenceCheckState(chart_vars=chart_vars,
-                                                exist_svs=[])
-        if chart_vars.event:
-          exist_cv.exist_event = utils.event_existence_for_place(
-              places[0], chart_vars.event, self.state.uttr.counters)
-          if not exist_cv.exist_event:
-            state.uttr.counters.err(
-                'failed_event_existence_check', {
-                    'places': places[:constants.DBG_LIST_LIMIT],
-                    'event': chart_vars.event
-                })
-        else:
+                                          extended_vars=None)
+
+      if not sv2extensions:
+        chart_vars_list = _build_chart_vars(state, sv, rank)
```

Comment:
(maybe for later) have more functions to wrap the logic in this file, or split this into multiple classes, files. As someone new to the code, it's a bit hard to follow the logic now..

Reply:
Splitting into multiple classes/files, SG.  Added TODO